### DISCUSSION TITLE
Cannot fetch log from Celery worker

### DISCUSSION THREAD

**Discussion by emredjan** [apache/airflow#26490] [General]
### Apache Airflow version

2.4.0

### What happened

When running tasks on a remote celery worker, webserver fails to fetch logs from the machine, giving a '403 - Forbidden' error on version 2.4.0. This behavior does not happen on 2.3.3, where the remote logs are retrieved and displayed successfully.
The `webserver / secret_key` configuration is the same in all nodes (the config files are synced), and their time is synchronized using a central NTP server, making the solution in the warning message not applicable.

My limited analysis pointed to the `serve_logs.py` file, and the flask request object that's passed to it, but couldn't find the root cause.

### What you think should happen instead

It should fetch and show remote celery worker logs on the webserver UI correctly, as it did in previous versions.

### How to reproduce

Use airflow version 2.4.0
Use CeleryExecutor with RabbitMQ
Use a separate Celery worker machine
Run a dag/task on the remote worker
Try to display task log on the web UI

### Operating System

Red Hat Enterprise Linux 8.6 (Ootpa)

### Versions of Apache Airflow Providers

```
apache-airflow-providers-celery==3.0.0
apache-airflow-providers-common-sql==1.1.0
apache-airflow-providers-ftp==3.0.0
apache-airflow-providers-hashicorp==3.0.0
apache-airflow-providers-http==3.0.0
apache-airflow-providers-imap==3.0.0
apache-airflow-providers-microsoft-mssql==3.0.0
apache-airflow-providers-mysql==3.0.0
apache-airflow-providers-odbc==3.0.0
apache-airflow-providers-sftp==3.0.0
apache-airflow-providers-sqlite==3.0.0
apache-airflow-providers-ssh==3.0.0
```

### Deployment

Virtualenv installation

### Deployment details

Using CeleryExecutor / rabbitmq with 2 servers

### Anything else

All remote task executions has the same problem.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
Could be connected to IPV6 issue #24846 - which has been fixed in Airflow 2.4.0 (relased today) see https://github.com/apache/airflow/issues/26201  - please upgrade to 2.4.0 and see if the problem persists (and please comment here). If it does not fix it (or even before)  - taje a look at the logs of webserver and they should contain error logs/explanation where it was from.

Converting to discussion as this is likely solved or need more detailed looking in the logs.

**Reply by emredjan** [reply]
Sorry if not clear, but this already happened in 2.4.0, which I installed today (It was fine on 2.3.3 and before).
Running on Python 3.10.5.

This is the message from the web UI:

![image](https://user-images.githubusercontent.com/389091/191034865-84bfef74-7639-45ce-9338-d51b9f80deac.png)

This is the relevant log from the remote celery worker when the web UI requests the task log:

```
Sep 19 15:55:05 airflow[3734318]: [2022-09-19 15:55:05,027: INFO/ForkPoolWorker-15] Task airflow.executors.celery_executor.execute_command[af1dec61-b437-4605-980c-de2d65f60bd7] succeeded in 0.9207695280201733s: None
Sep 19 15:55:11 airflow[3734281]: [2022-09-19 15:55:11,177] {serve_logs.py:104} WARNING - Unknown error
Sep 19 15:55:11 airflow[3734281]: Traceback (most recent call last):
Sep 19 15:55:11 airflow[3734281]:   File "/data/apps/.pyenv/versions/3.10.5/envs/airflow-py310/lib/python3.10/site-packages/airflow/utils/serve_logs.py", line 64, in validate_pre_signed_url
Sep 19 15:55:11 airflow[3734281]:     request_filename = request.view_args['filename']
Sep 19 15:55:11 airflow[3734281]: TypeError: 'NoneType' object is not subscriptable
```

Looks like the `request.view_args` is `None` in the validation function. Don't know whether it's the root cause though.

Do you need any additional logs?

**Comment by patricker** [comment]
We are seeing the same exact error in our worker logs coming from `serve_logs.py` during upgrade testing from 2.3.4 to 2.4.0.
Python 3.9, Celery, Kubernetes.

**Comment by patricker** [comment]
I think this was resolved on https://github.com/apache/airflow/issues/26492?

**Comment by potiuk** [comment]
Yes. There was a bug in 2.4.0 that has a workaround you can apply manually from https://github.com/apache/airflow/pull/26493  and it will be fixed permanently in 2.4.1

**Reply by HassanAlahmed** [reply]
Hi @potiuk - I'm trying to apply the workaround manually. Our airflow runs in Azure Kubernetes Cluster, so I fixed file_task_handler.py that's in the worker and webserver pods. However, I'm still receiving the same error:

```
9/26/2022, 1:23:04 PM
fruits360-worker-0
9c71a1869b43633e1cec6e6a6d76d5dec1fe20b4535d8ea79d57e9271f54a68c
[2022-09-26T10:23:04.907+0000] {serve_logs.py:104} WARNING - Unknown error
9/26/2022, 1:23:04 PM
fruits360-worker-0
9c71a1869b43633e1cec6e6a6d76d5dec1fe20b4535d8ea79d57e9271f54a68c
Traceback (most recent call last):
9/26/2022, 1:23:04 PM
fruits360-worker-0
9c71a1869b43633e1cec6e6a6d76d5dec1fe20b4535d8ea79d57e9271f54a68c
File "/home/airflow/.local/lib/python3.7/site-packages/airflow/utils/serve_logs.py", line 64, in validate_pre_signed_url
9/26/2022, 1:23:04 PM
fruits360-worker-0
9c71a1869b43633e1cec6e6a6d76d5dec1fe20b4535d8ea79d57e9271f54a68c
request_filename = request.view_args['filename']
9/26/2022, 1:23:04 PM
fruits360-worker-0
9c71a1869b43633e1cec6e6a6d76d5dec1fe20b4535d8ea79d57e9271f54a68c
TypeError: 'NoneType' object is not subscriptable
```

I'm not sure what you mean by "a workaround you can apply manually", but was my way wrong?

**Reply by potiuk** [reply]
If you use K8S cluster, you have to apply it to the code in the image you run in your Cluster - basically it means you need to build and use a custom image with the patch in (until we release 2.4.1).

**Reply by vinitpayal** [reply]
Thanks @potiuk we were facing the same issue but your fix at https://github.com/apache/airflow/pull/26493 made it work.

**Reply by potiuk** [reply]
It was my bug too, so I am not sure if I can take a credit for that one :D

**Reply by HassanAlahmed** [reply]
@potiuk We all make them mistakes. You found and fixed the issue promptly, so of course you can take credit.

So I checked tag 2.4.0 from airflow, add the bug fix, and built and image using `docker build --tag <tag>`. When I used the image in the helm chart deployment, the run-migration-job failed with the following error:
```
alembic.util.exc.CommandError: Can't locate revision identified by 'ecb43d2a1842'
```

I checked the `alembic_version` table in my metadata db, and `ecb43d2a1842` is already applied. I'm not sure why the run-migration-job is  trying to apply this migration, and I'm not sure why the revision cannot be located. Could you please shed some light on this? Sorry for bothering you so much with my questions. I really appreciate your help.

**Reply by potiuk** [reply]
I think the problem is that for some reson your image is not based on airflow 2.4.0. That would be the error. There might be multiple reasons for that - wrong tag, wrong docker image, accidental downgrade of airflow by some unrelated dependency. I think you should verify if your Docker image really contains the right version and look closely that the output of your docker build.

**Reply by potiuk** [reply]
(ultimately I think waiting for 2.4.1 - rc planned today or tomorrow) is probably a good idea :)

**Reply by potiuk** [reply]
2.4.1 is out with the fix :)
